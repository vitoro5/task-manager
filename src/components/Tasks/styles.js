import styled from "styled-components";

export const Container = styled.ul`
  list-style: none;
  display: flex;
  flex-direction: column;

  li {
    button {
      display: flex;
      flex-direction: row;
      align-items: center;
      color: #363636;
      border: none;
      background: transparent;
      font-family: "Mali", cursive;
      cursor: pointer;

      svg {
        color: #dbdbdb;
      }

      p {
        margin-left: 10px;
      }
    }
  }
`;
