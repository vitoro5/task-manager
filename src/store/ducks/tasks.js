export const Types = {
  ADD_TASK: "tasks/ADD_TASK",
  TOGGLE_TASK: "tasks/TOGGLE_TASK",
  CLEAR_TASKS: "tasks/CLEAR_TASKS"
};

const INITIAL_STATE = {
  data: []
};

export default function tasks(state = INITIAL_STATE, action) {
  switch (action.type) {
    case Types.ADD_TASK:
      return {
        ...state,
        data: [...state.data, action.task]
      };

    case Types.TOGGLE_TASK:
      const index = state.data.indexOf(action.task);

      const newTasks = state.data.map((task, i) => {
        if (i !== index) {
          return task;
        }

        const updatedTask = () => {
          return { ...task, completed: !task.completed };
        };

        return updatedTask();
      });
      return { ...state, data: newTasks };

    case Types.CLEAR_TASKS:
      return {
        ...state,
        data: []
      };

    default:
      return state;
  }
}

export const Creators = {
  addTask: task => ({
    type: Types.ADD_TASK,
    task
  }),

  toggleTask: task => ({
    type: Types.TOGGLE_TASK,
    task
  }),

  clearTasks: () => ({
    type: Types.CLEAR_TASKS
  })
};
