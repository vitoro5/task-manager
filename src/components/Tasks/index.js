import React from "react";
import { useSelector, useDispatch } from "react-redux";

import { Creators } from "../../store/ducks/tasks";

import { Container } from "./styles";

import { ReactComponent as Unchecked } from "../../icons/circle.svg";
import { ReactComponent as Checked } from "../../icons/check-mark.svg";

function Tasks({ filter }) {
  const tasks = useSelector(({ tasks }) => tasks.data);
  const dispatch = useDispatch();

  const handleClick = task => {
    dispatch(Creators.toggleTask(task));
  };

  if (filter === "all") {
    return (
      <Container>
        {tasks.map((task, i) => (
          <li key={i}>
            <button onClick={() => handleClick(task)}>
              {!task.completed && <Unchecked width="12" height="12" />}
              {task.completed && <Checked width="12" height="12" />}
              <p>{task.value}</p>
            </button>
          </li>
        ))}
      </Container>
    );
  } else {
    return (
      <Container>
        {tasks
          .filter(task => task.completed === filter)
          .map((task, i) => (
            <li key={i}>
              <button onClick={() => handleClick(task)}>
                {!task.completed && <Unchecked width="12" height="12" />}
                {task.completed && <Checked width="12" height="12" />}
                <p>{task.value}</p>
              </button>
            </li>
          ))}
      </Container>
    );
  }
}

export default Tasks;
