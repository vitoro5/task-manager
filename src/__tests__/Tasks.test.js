import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { createStore } from "redux";

import Tasks from "../components/Tasks";
import reducer from "../store/ducks";
import { Creators as Actions } from "../store/ducks/tasks";

describe("<Tasks /> unit test", () => {
  const initialState = {
    tasks: {
      data: [
        { value: "testing1", completed: false },
        { value: "testing2", completed: false }
      ],
      filter: "all"
    }
  };

  const getWrapper = (mockStore = createStore(reducer, initialState)) =>
    mount(
      <Provider store={mockStore}>
        <Tasks />
      </Provider>
    );

  it("should renders without crashing", () => {
    getWrapper();
  });

  it("should renders when filter is setted to all", () => {
    const wrapper = getWrapper();

    expect(wrapper).toMatchSnapshot();
  });

  it("should renders when filter is not setted to all", () => {
    const newState = {
      tasks: {
        data: [{ value: "testing", completed: false }],
        filter: true
      }
    };

    const mockStore = createStore(reducer, newState);

    const wrapper = getWrapper(mockStore);
    expect(wrapper).toMatchSnapshot();
  });

  it("should toggle completed status on click", () => {
    const mockStore = createStore(reducer, initialState);
    mockStore.dispatch = jest.fn();

    const wrapper = getWrapper(mockStore);
    let button = wrapper.find("button").first();

    button.simulate("click");
    expect(mockStore.dispatch).toHaveBeenCalledWith(
      Actions.toggleTask(initialState.tasks.data[0])
    );
  });
});
