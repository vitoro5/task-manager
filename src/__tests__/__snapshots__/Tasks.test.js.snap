// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tasks /> unit test should renders when filter is not setted to all 1`] = `
.c0 {
  list-style: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 li button {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #363636;
  border: none;
  background: transparent;
  font-family: "Mali",cursive;
  cursor: pointer;
}

.c0 li button svg {
  color: #dbdbdb;
}

.c0 li button p {
  margin-left: 10px;
}

<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Tasks>
    <styled.ul>
      <ul
        className="c0"
      />
    </styled.ul>
  </Tasks>
</Provider>
`;

exports[`<Tasks /> unit test should renders when filter is setted to all 1`] = `
.c0 {
  list-style: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 li button {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #363636;
  border: none;
  background: transparent;
  font-family: "Mali",cursive;
  cursor: pointer;
}

.c0 li button svg {
  color: #dbdbdb;
}

.c0 li button p {
  margin-left: 10px;
}

<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Tasks>
    <styled.ul>
      <ul
        className="c0"
      >
        <li
          key="0"
        >
          <button
            onClick={[Function]}
          >
            <ForwardRef(SvgCircle)
              height="12"
              width="12"
            >
              <svg
                height="12"
                width="12"
              >
                circle.svg
              </svg>
            </ForwardRef(SvgCircle)>
            <p>
              testing1
            </p>
          </button>
        </li>
        <li
          key="1"
        >
          <button
            onClick={[Function]}
          >
            <ForwardRef(SvgCircle)
              height="12"
              width="12"
            >
              <svg
                height="12"
                width="12"
              >
                circle.svg
              </svg>
            </ForwardRef(SvgCircle)>
            <p>
              testing2
            </p>
          </button>
        </li>
      </ul>
    </styled.ul>
  </Tasks>
</Provider>
`;
